rust/src/measurements/make_private_lazyframe/select/mod.rs

```diff
@@ -0,0 +1,132 @@
+use std::collections::BTreeSet;
+
+use crate::combinators::{make_basic_composition, BasicCompositionMeasure};
+use crate::core::{Function, Measurement, MetricSpace, StabilityMap, Transformation};
+use crate::domains::{ExprContext, ExprDomain, LogicalPlanDomain, MarginPub};
+use crate::error::*;
+use crate::measurements::make_private_expr;
+use crate::metrics::PartitionDistance;
+use crate::transformations::traits::UnboundedMetric;
+use crate::transformations::{DatasetMetric, StableLogicalPlan};
+use make_private_expr::PrivateExpr;
+use polars::prelude::*;
+
+#[cfg(test)]
+mod test;
+
+/// Create a private version of an aggregate operation on a LazyFrame.
+///
+/// # Arguments
+/// * `input_domain` - The domain of the input LazyFrame.
+/// * `input_metric` - The metric of the input LazyFrame.
+/// * `output_measure` - The measure of the output LazyFrame.
+/// * `plan` - The LazyFrame to transform.
+/// * `global_scale` - The parameter for the measurement.
+pub fn make_private_select<MS, MI, MO>(
+    input_domain: LogicalPlanDomain,
+    input_metric: MS,
+    output_measure: MO,
+    plan: LogicalPlan,
+    global_scale: Option<f64>,
+) -> Fallible<Measurement<LogicalPlanDomain, LogicalPlan, MS, MO>>
+where
+    MS: 'static + DatasetMetric,
+    MI: 'static + UnboundedMetric,
+    MO: 'static + BasicCompositionMeasure,
+    Expr: PrivateExpr<PartitionDistance<MI>, MO>,
+    LogicalPlan: StableLogicalPlan<MS, MI>,
+    (LogicalPlanDomain, MS): MetricSpace,
+    (LogicalPlanDomain, MI): MetricSpace,
+    (ExprDomain, MI): MetricSpace,
+    (ExprDomain, PartitionDistance<MI>): MetricSpace,
+{
+    let LogicalPlan::Projection { expr, input, .. } = plan.clone() else {
+        return fallible!(MakeMeasurement, "Expected selection in logical plan");
```

Comment:
Not feasible to just have stronger typing on the function, instead of using a runtime error?

Reply:
From the compiler's point of view, all it knows is that the type is a LogicalPlan enum. The type system does not distinguish the variant of the enum-- that is only known at runtime. It would have been nice if each variant were a tuple struct holding a separate struct (like `LogicalPlan::Projection(logical_plan_projection)`) instead of being an enum variant struct (like `LogicalPlan::Projection { expr, input, .. }`), but it's not our code so we have to match here as well.

Reply:
Thanks for the explanation.